pkgs/applications/virtualization/singularity/template.nix

```diff
@@ -0,0 +1,163 @@
+{ lib
+, util-linux
+, gpgme
+, openssl
+, libuuid
+, bash
+, coreutils
+, which
+, makeWrapper
+, cryptsetup
+, squashfsTools
+, nvidia-docker # For nvidia-container-cli
+, glibc # For ldconfig needed by nvidia-container-cli
+, go
+, buildGoModule
+, pname
+, version
+, src
+, projectName # "apptainer" or "singularity"
+  # These cannot be correctly override with `overrideAttrs`
+, vendorSha256 ? null
+, deleteVendor ? false
+, proxyVendor ? false
+, doCheck ? true
+, withoutSuid ? true # Required to run containers without suid
+, enableNvidiaContainerCli ? true
+}:
+
+with lib;
+let
+  propagatedBuildInputs = [ coreutils squashfsTools bash ];
+in
+buildGoModule {
+  inherit pname version src;
+
+  # `buildGoModule` reads the `go.mod` file,
+  # so no need to specify `goPackagePath`
+
+  # All three projects provide vendored source tarball
+  # so we set vendorSha256 = null;
+  # One who use unvendored source can override this
+  # with `override`
+  inherit vendorSha256 deleteVendor proxyVendor;
+
+  # apptainer and singularity-ce both want to reference Go
+  # to compile the extensions when building container images
+  allowGoReference = true;
+
+  passthru = {
+    inherit projectName;
+  };
+
+  buildInputs = [ gpgme openssl libuuid ];
+  nativeBuildInputs = [ util-linux which makeWrapper cryptsetup ];
+  inherit propagatedBuildInputs;
+
+  outputs = [ "out" "man" ];
+
+  postPatch = ''
+    substituteInPlace internal/pkg/build/files/copy.go \
+      --replace /bin/cp ${coreutils}/bin/cp
+  '';
+
+  configureScript = "./mconfig";
+
+  configureFlags = [
+    "--localstatedir=/var"
+  ]
+  ++ lib.optional withoutSuid "--without-suid";
+
+  # Using `buildGoModule` instead of `buildGoPackage`,
+  # We no longer need to `cd go/src/${goPackagePath}`
+  postConfigure = ''
+    patchShebangs .
+
+    # Patching the hard-coded [Dd]efaultPath
+    for subPath in cmd/internal/cli/actions.go internal/pkg/util/env/env.go e2e/env/env.go; do
+      if [ -f "$subPath" ]; then
+        echo "Patching [Dd]efaultPath in \"$subPath\""
+        sed -i 's|^\([ \t]*[Dd]efaultPath[ \t]*[:]\?=[ \t]*\)"[^"]*"|\1"${lib.makeBinPath propagatedBuildInputs}"|' "$subPath"
+      fi
+    done
+
+    # Try to make use of the existing configureFlags and configureFlagsArray
+    # Code borrowed from pkgs/stdenv/generic/setup.sh configurePhase()
```

Comment:
Hmm, was this meant to be rewritten to reuse existing `configurePhase` later before merging?

Reply:
`buildGoModule` rewrites the `configurePhase` to do the module vendoring before the build starts.

This PR re-introduce the functionality of the default configurePhase by copying the related code from [`pkgs/stdenv/generic/setup.sh`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L1006-L1061)

Reply:
Resolved per https://github.com/NixOS/nixpkgs/pull/158486#discussion_r901013477